### PR TITLE
Improve Efficiency of SnapshotsService CS Apply

### PR DESCRIPTION
**Pull request by original-brownbear**
This change removes the redundant submitting of two separate cluster state updates
for the node configuration changes and routing changes that affect snapshots.
Since we submitted the task to deal with node configuration changes every time on master
fail-over we could also move the BwC cleanup loop that removes `INIT` state snapshots as well
as snapshots that have all their shards completed into this cluster state update task.

Aside from improving efficiency overall this change has the fortunate side effect of moving
all snapshot finalization to the CS update thread. This is helpful for concurrent snapshots
since it makes it very natural and straight forward to order snapshot finalizations by exploiting
that they are all initiated on the same thread.

### CODE REVIEW

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 1]
```diff
@@ -678,11 +633,26 @@ private static boolean waitingShardsStartedOrUnassigned(SnapshotsInProgress snap
     }
 
     private static boolean removedNodesCleanupNeeded(SnapshotsInProgress snapshotsInProgress, List<DiscoveryNode> removedNodes) {
-        // If at least one shard was running on a removed node - we need to fail it
-        return removedNodes.isEmpty() == false && snapshotsInProgress.entries().stream().flatMap(snapshot ->
-                StreamSupport.stream(((Iterable<ShardSnapshotStatus>) () -> snapshot.shards().valuesIt()).spliterator(), false)
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:683]: This was really evil for large snapshots (even more so in the context of concurrent snapshots). We would iterate all the shards even if no nodes were removed or the snapshot finished already.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 2]
```diff
@@ -697,12 +667,6 @@ private void endSnapshot(SnapshotsInProgress.Entry entry, Metadata metadata) {
             return;
         }
         final Snapshot snapshot = entry.snapshot();
-        if (entry.repositoryStateId() == RepositoryData.UNKNOWN_REPO_GEN) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:700]: This is unnecessary now, we will not pass a snapshot with `UNKNOWN_REPO_GEN` to this method any longer, we just remove it from the cluster state in `processWaitingShardsAndRemovedNodes` and are done with it.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 3]
```diff
@@ -1424,7 +1388,8 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
                             if (endingSnapshots.contains(request.snapshot()) == false) {
                                 final SnapshotsInProgress snapshotsInProgress = newState.custom(SnapshotsInProgress.TYPE);
                                 final SnapshotsInProgress.Entry updatedEntry = snapshotsInProgress.snapshot(request.snapshot());
-                                if (updatedEntry.state().completed()) {
+                                // If the entry is still in the cluster state and is completed, try finalizing the snapshot in the repo
+                                if (updatedEntry != null && updatedEntry.state().completed()) {
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:1392]: This was a bug introduced in #56669 , we could NPE here if a data node were to resend a state update for an already finished snapshot (master fail-over or node disconnect can easily lead to this) causing an NPE here needlessly.

File: server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java [hunk 4]
```diff
@@ -616,8 +560,8 @@ public void clusterStateProcessed(String source, ClusterState oldState, ClusterS
         });
     }
 
-    private static ImmutableOpenMap<ShardId, ShardSnapshotStatus> processWaitingShards(
-            ImmutableOpenMap<ShardId, ShardSnapshotStatus> snapshotShards, RoutingTable routingTable) {
+    private static ImmutableOpenMap<ShardId, ShardSnapshotStatus> processWaitingShardsAndRemovedNodes(
```
Comment (original-brownbear) [inline server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java:563]: Combining both node removal and shard starting into one method saves us iterating all the shards twice for the relevant updates to each snapshot.

**Comment by original-brownbear** [comment]
Jenkins run elasticsearch-ci/1

**Review by ywelsch — APPROVED** [review]
LGTM

**Comment by original-brownbear** [comment]
Thanks Yannick!

**Pull request merged**